drivers/flash/flash_stm32wbax.c

```diff
@@ -0,0 +1,309 @@
+
+	/* if the non-secure control register is locked, do not fail silently */
+	if (regs->NSCR & FLASH_STM32_NSLOCK) {
+		LOG_ERR("NSCR locked\n");
+		return -EIO;
+	}
+
+	/* Check that no Flash main memory operation is ongoing */
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	/* Check if this double word is erased */
+	if ((flash[0] != 0xFFFFFFFFUL) || (flash[1] != 0xFFFFFFFFUL) ||
+		(flash[2] != 0xFFFFFFFFUL) || (flash[3] != 0xFFFFFFFFUL)) {
+		LOG_ERR("Word at offs %ld not erased", (long)offset);
+		return -EIO;
+	}
+
+	/* Set the NSPG bit */
+	regs->NSCR |= FLASH_STM32_NSPG;
+
+	/* Flush the register write */
+	tmp = regs->NSCR;
+
+	/* Perform the data write operation at the desired memory address */
+	flash[0] = buff[0];
+	flash[1] = buff[1];
+	flash[2] = buff[2];
+	flash[3] = buff[3];
+
+	/* Wait until the NSBSY bit is cleared */
+	rc = flash_stm32_wait_flash_idle(dev);
+
+	/* Clear the NSPG bit */
+	regs->NSCR &= (~FLASH_STM32_NSPG);
+
+	return rc;
+}
+
+static int erase_page(const struct device *dev, unsigned int offset)
+{
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	uint32_t tmp;
+	int rc;
+	int page;
+
+	/* if the non-secure control register is locked,do not fail silently */
+	if (regs->NSCR & FLASH_STM32_NSLOCK) {
+		LOG_ERR("NSCR locked\n");
+		return -EIO;
+	}
+
+	/* Check that no Flash memory operation is ongoing */
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	page = offset / FLASH_PAGE_SIZE;
+	LOG_DBG("Erase page %d\n", page);
+
+	/* Set the NSPER bit and select the page you wish to erase */
+	regs->NSCR |= FLASH_STM32_NSPER;
+	regs->NSCR &= ~FLASH_STM32_NSPNB_MSK;
+	regs->NSCR |= (page << FLASH_STM32_NSPNB_POS);
+
+	/* Set the NSSTRT bit */
+	regs->NSCR |= FLASH_STM32_NSSTRT;
+
+	/* flush the register write */
+	tmp = regs->NSCR;
+
+	/* Wait for the NSBSY bit */
+	rc = flash_stm32_wait_flash_idle(dev);
+
+	regs->NSCR &= ~(FLASH_STM32_NSPER);
+
+	return rc;
+}
+
+int flash_stm32_block_erase_loop(const struct device *dev,
+				 unsigned int offset,
+				 unsigned int len)
+{
+	unsigned int address = offset;
+	int rc = 0;
+	bool icache_enabled = LL_ICACHE_IsEnabled();
+
+	if (icache_enabled) {
+		/* Disable icache, this will start the invalidation procedure.
+		 * All changes(erase/write) to flash memory should happen when
+		 * i-cache is disabled. A write to flash performed without
+		 * disabling i-cache will set ERRF error flag in SR register.
+		 */
+		rc = stm32_icache_disable();
+		if (rc != 0) {
+			return rc;
+		}
+	}
+
+	for (; address <= offset + len - 1 ; address += FLASH_PAGE_SIZE) {
+		rc = erase_page(dev, address);
+		if (rc < 0) {
+			break;
+		}
+	}
+
+	if (icache_enabled) {
+		/* Since i-cache was disabled, this would start the
+		 * invalidation procedure, so wait for completion.
+		 */
+		rc = icache_wait_for_invalidate_complete();
+
+		/* I-cache should be enabled only after the
+		 * invalidation is complete.
+		 */
+		stm32_icache_enable();
```

Comment:
should the cache be enabled if rc != 0 (timeout for examples ?)

Reply:
Let's keep it that way for now.